src/devices/machine/vrender0.cpp

```diff
@@ -422,66 +428,108 @@ WRITE16_MEMBER(vrender0soc_device::tmcnt_w)
 // in a work RAM to area $4500000 transfer, probably to extend something ...
-inline int vrender0soc_device::dma_setup_hold(uint8_t setting, uint8_t bitmask)
+static inline int dma_setup_hold_dec(u8 setting, u8 bitmask, u8 decmask)
+{
+	const int inc = setting & bitmask ? 0 : (setting & 2) ? 4 : (1 << (setting & 1));
+	return setting & decmask ? -inc : inc;
+}
+
+template<int Which> u32 vrender0soc_device::dmasa_r() { return m_dma[Which].src; }
+template<int Which> void vrender0soc_device::dmasa_w(offs_t offset, u32 data, u32 mem_mask)
 {
-	return setting & bitmask ? 0 : (setting & 2) ? 4 : (1 << (setting & 1));
+	COMBINE_DATA(&m_dma[Which].src);
+	if (m_dma[Which].ctrl & DMA_INTERNAL_WRITE) // write internal counter
+		m_dma[Which].int_src = m_dma[Which].src;
 }
 
-template<int Which> READ32_MEMBER(vrender0soc_device::dmasa_r) { return m_dma[Which].src; }
-template<int Which> WRITE32_MEMBER(vrender0soc_device::dmasa_w) { COMBINE_DATA(&m_dma[Which].src); }
-template<int Which> READ32_MEMBER(vrender0soc_device::dmada_r) { return m_dma[Which].dst; }
-template<int Which> WRITE32_MEMBER(vrender0soc_device::dmada_w) { COMBINE_DATA(&m_dma[Which].dst); }
-template<int Which> READ32_MEMBER(vrender0soc_device::dmatc_r) { return m_dma[Which].size; }
-template<int Which> WRITE32_MEMBER(vrender0soc_device::dmatc_w) { COMBINE_DATA(&m_dma[Which].size); }
-template<int Which> READ32_MEMBER(vrender0soc_device::dmac_r) { return m_dma[Which].ctrl; }
+template<int Which> u32 vrender0soc_device::dmada_r() { return m_dma[Which].dst; }
+template<int Which> void vrender0soc_device::dmada_w(offs_t offset, u32 data, u32 mem_mask)
+{
+	COMBINE_DATA(&m_dma[Which].dst);
+	if (m_dma[Which].ctrl & DMA_INTERNAL_WRITE) // write internal counter
+		m_dma[Which].int_dst = m_dma[Which].dst;
+}
+
+template<int Which> u32 vrender0soc_device::dmatc_r() { return m_dma[Which].size; }
+template<int Which> void vrender0soc_device::dmatc_w(offs_t offset, u32 data, u32 mem_mask)
+{
+	mem_mask &= 0xffffff; // 24 bit
+	COMBINE_DATA(&m_dma[Which].size);
+	m_dma[Which].int_cnt = m_dma[Which].size; // write internal counter?
+}
+
+template<int Which> u32 vrender0soc_device::dmac_r() { return m_dma[Which].ctrl; }
 template<int Which>
-WRITE32_MEMBER(vrender0soc_device::dmac_w)
+void vrender0soc_device::dmac_w(offs_t offset, u32 data, u32 mem_mask)
 {
-	if (((data ^ m_dma[Which].ctrl) & (1 << 10)) && (data & (1 << 10)))   //DMAOn
+	const u32 old = m_dma[Which].ctrl;
+	data = COMBINE_DATA(&m_dma[Which].ctrl);
+	if ((!(old & DMA_ENABLE)) && (data & DMA_ENABLE))   //DMAOn
 	{
-		uint32_t const CTR = data;
-		uint32_t const SRC = m_dma[Which].src;
-		uint32_t const DST = m_dma[Which].dst;
-		uint32_t const CNT = m_dma[Which].size;
-		const int src_inc = dma_setup_hold(CTR, 0x20);
-		const int dst_inc = dma_setup_hold(CTR, 0x08);
-
-		if ((CTR & 0xd4) != 0)
-			popmessage("DMA%d with unhandled mode %02x, contact MAMEdev",Which,CTR);
```

Comment:
So do you have a program that actually tests all these DMA modes?
If not you're (again) coding in blind faith, which is (again) against our policies (there's a popmessage for a reason).

Reply:
if it's based on documentation, which cam seems to have link, I wouldn't be entirely against it, maybe add a popmessage if it's used, that way, if a use case is found, and the code works, the message can be removed.

Reply:
popmessage is still here, but I fix comments when unhandled mode accessed.